### PR TITLE
Conti/emit tested integration versions

### PR DESCRIPTION
**Pull request by wconti27**
Support receiving and emitting tested integrations and their versions. This data will be sent to Metabase to create dashboards for each tracer.

### CODE REVIEW

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]

File: ddapm_test_agent/agent.py [hunk 1]
```diff
@@ -77,6 +78,17 @@ def _parse_csv(s: str) -> List[str]:
     return [s.strip() for s in s.split(",") if s.strip() != ""]
 
 
+def is_semver_and_release_version(version: str) -> bool:
+    """Attempts to parse the inputted version and returns whether the version
+    is a valid semver AND is not a pre-release or dev version.
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddapm_test_agent/agent.py]: Changed this as spoken about

File: ddapm_test_agent/agent.py [hunk 2]
```diff
@@ -188,6 +199,9 @@ def __init__(self):
         self._trace_check_results_by_check: Dict[str, Dict[str, Dict[str, int]]] = defaultdict(
             default_value_trace_check_results_by_check
         )
+        self._integrations: Dict[str, dict] = {}
+        self._sent_integration_versions = set()
+        self._tracer_version = None
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddapm_test_agent/agent.py]: I changed this code, no longer keeping a list of seen integrations, instead grabbing integration requests from stored requests. This should be set

File: ddapm_test_agent/agent.py [hunk 3]
```diff
@@ -142,6 +141,8 @@ async def _prepare_and_send_request(data, request, headers):
     agent_url = request.app["agent_url"]
     full_agent_url = agent_url + request.path
     log.info("Forwarding request to agent at %r", full_agent_url)
+    if "DD_API_KEY" not in headers:
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:144]: why was this added? it's not used anywhere?

File: ddapm_test_agent/agent.py [hunk 4]
```diff
@@ -387,6 +388,60 @@ async def _tracestats_by_session(self, token: Optional[str]) -> List[v06StatsPay
...
+                    )
+                    req["tracer_version"] = data.get("tracer_version", None)
+                    req["tracer_language"] = data.get("tracer_language", None)
+                    integration_requests.append(req)
+                elif include_sent_integrations:
+                    integration_requests.append(req)
+            # check if integration data was provided in the trace request instead
+            elif (
+                "_dd_trace_env_variables" in req
+                and "DD_PLUGIN" in req["_dd_trace_env_variables"]
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:417]: since we're referring to them as "integrations" in the test agent, does it make sense to use `DD_INTEGRATION` as the env var?

File: ddapm_test_agent/agent.py [hunk 5]
```diff
@@ -459,6 +514,82 @@ async def handle_v2_apmtelemetry(self, request: Request) -> web.Response:
         # TODO: Snapshots
         return web.HTTPOk()
 
+    async def handle_put_tested_integrations(self, request: Request) -> web.Response:
+        # we need to store the request manually since this is not a real DD agent endpoint
+        await self._store_request(request)
+        await self.update_seen_integration_versions()
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:520]: An alternate solution for this would be to add an endpoint that dumps all the seen integrations to a specified file. That way the path can be customized if need-be.

This way we aren't writing to disk on every single request.

File: ddapm_test_agent/agent.py [hunk 6]
```diff
@@ -898,13 +1032,15 @@ def make_app(
             web.post("/test/session/responses/config", agent.handle_v07_remoteconfig_create),
             web.post("/test/session/responses/config/path", agent.handle_v07_remoteconfig_path_create),
             web.put("/test/session/responses/config", agent.handle_v07_remoteconfig_put),
+            web.put("/test/session/integrations", agent.handle_put_tested_integrations),
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:1035]: these endpoints need to be documented in the README, especially with the saving-seen-integrations-to-file functionality that exists that writes the integrations to some magic file.

Comment (Kyle-Verhoog) [reply ddapm_test_agent/agent.py:1035]: also nice to have some tests

Comment (Kyle-Verhoog) [reply ddapm_test_agent/agent.py:1035]: also a release note

File: ddapm_test_agent/agent.py (hunk 3 above)
Comment (wconti27) [reply ddapm_test_agent/agent.py:143]: removed, cant remember tbh

File: ddapm_test_agent/agent.py [hunk 7]
```diff
@@ -459,6 +509,53 @@ async def handle_v2_apmtelemetry(self, request: Request) -> web.Response:
...
+
+    async def handle_get_tested_integrations(self, request: Request) -> web.Response:
+        """Return all tested integrations according to integration data received by agent."""
+        text_headers = ["language_name", "tracer_version", "integration_name", "integration_version", "dependency_name"]
+        aggregated_text = ""
+        seen_integrations = set()
+        req_headers = {}
+
+        # get all requests associated with an integration
+        reqs = await self._integration_requests_by_session(token=None, include_sent_integrations=True)
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:525]:
```suggestion
        reqs = await self._integration_requests_by_session(token=_session_token(request), include_sent_integrations=True)
```

**Review by Kyle-Verhoog — APPROVED** [review]
lgtm!
